py3status/modules/speedtest.py

```diff
@@ -0,0 +1,296 @@
+
+@author Cyril Levis (@cyrinux)
+
+Examples:
+```
+# colored based on thresholds
+speedtest {
+    thresholds = {
+        "download": [(0, "bad"), (1024, "degraded"), (1024 * 1024, "good")],
+        "ping": [(200, "bad"), (150, "orange"), (100, "degraded"), (10, "good")],
+        "quality": [
+            ("ok", "good"),
+            ("bad", "degraded"),
+            ("faster", "good"),
+            ("slower", "degraded"),
+            ("faster", "good"),
+        ],
+        "upload": [(0, "bad"), (1024, "degraded"), (1024 * 1024, "good")],
+    }
+    format = '[\?color=ping {ping} ms] [\?color=upload Up {upload} {upload_unit}]'
+    format += ' [\?color=download Down {download} {download_unit}]'
+}
+
+# colored based on comparation between two last runs
+speedtest {
+    format = '[[\?if=download_raw<previous_download_raw&color=degraded Faster]'
+    format += '|[\?color=ok Slower]]'
+}
+
+# compare only download
+speedtest {
+    format = '[[\?if=quality=slower&color=download_raw ↓ {download} {download_unit}]'
+    format += '[\?if=quality=faster&color=download_raw ↑ {download} {download_unit}]'
+    format += '[\?if=quality=ok&color=download_raw -> {download} {download_unit}]'
+    format += '[\?if=quality=bad&color=download_raw x {download} {download_unit}]]'
+    format += '|speedtest'
+}
+```
+
+SAMPLE OUTPUT
+{'full_text': u'Speedtest Down 0.247 MB/s Up 0.453 MB/s'}
+"""
+
+from json import loads
+from numbers import Number
+
+STRING_NOT_INSTALLED = "not installed"
+MISSING_PLACEHOLDER = "{download} or {upload} placeholder required"
+
+
+class Py3status:
+    """
+    """
+
+    # available configuration parameters
+    button_refresh = 2
+    button_share = None
+    format = (
+        u"Speedtest [Up {upload} {upload_unit}] [Down {download} {download_unit}]"
+    )
+    server_id = None
+    si_units = False
+    sleep_timeout = 5
+    thresholds = {
+        "download": [(0, "bad"), (1024, "degraded"), (1024 * 1024, "good")],
+        "ping": [(200, "bad"), (150, "orange"), (100, "degraded"), (10, "good")],
+        "quality": [
+            ("ok", "good"),
+            ("bad", "degraded"),
+            ("faster", "good"),
+            ("slower", "degraded"),
+            ("faster", "good"),
+        ],
+        "upload": [(0, "bad"), (1024, "degraded"), (1024 * 1024, "good")],
+    }
+    timeout = 10
+    unit_bitrate = "MB/s"
+    unit_size = "MB"
+
+    def post_config_hook(self):
+        if not self.py3.check_commands("speedtest-cli"):
+            raise Exception(STRING_NOT_INSTALLED)
+
+        self.first_run = True
+
+        self.placeholders = self.py3.get_placeholders_list(self.format)
+
+        self.command = ["speedtest-cli --json --secure"]
+
+        # if download* or upload* missing, run complete test
+        if any(
+            key.startswith(x)
+            for x in ["download", "upload"]
+            for key in self.placeholders
```

Comment:
We should remove this over-complicated if-statement. `post_config_hook` is run only once.

Reply:
Do you need all three `ifs` or just two `ifs`?

Reply:
Idk, for the moment i see how get what i want with 3 ... I will reread...

Reply:
3